### DISCUSSION TITLE
AWS / EC2 - arm64 linux centos image - /lib64/libm.so.6: version `GLIBC_2.27' not found

### DISCUSSION THREAD

**Discussion by johndpope** [PostgREST/postgrest#2012] [Q&A] [answered]
### Environment

* PostgreSQL version: (if using docker, specify the image)
* latest
* PostgREST version: (if using docker, specify the image)
* latest
* Operating system:

### Description of issue
can't run any of the released images on arm64 hardware. 
I tried many variations and previous builds from users but not working. 

./postgrest: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by ./postgrest)

UPDATE - it says it has 2.26 

```shell

➜  ~ sudo yum install glibc-devel
Loaded plugins: extras_suggestions, langpacks, priorities, update-motd
amzn2-core                                                                                                 | 3.7 kB  00:00:00
200 packages excluded due to repository priority protections
Package glibc-devel-2.26-55.amzn2.aarch64 already installed and latest version
Nothing to do
```

https://aws.amazon.com/amazon-linux-2/release-notes/

**glibc-2.26-55.amzn2.aarch64
glibc-2.26-55.amzn2.x86_64
glibc-all-langpacks-2.26-55.amzn2.aarch64
glibc-all-langpacks-2.26-55.amzn2.x86_64
glibc-common-2.26-55.amzn2.aarch64
glibc-common-2.26-55.amzn2.x86_64
glibc-devel-2.26-55.amzn2.x86_64
glibc-headers-2.26-55.amzn2.x86_64
glibc-langpack-en-2.26-55.amzn2.aarch64
glibc-langpack-en-2.26-55.amzn2.x86_64
glibc-locale-source-2.26-55.amzn2.aarch64
glibc-locale-source-2.26-55.amzn2.x86_64
glibc-minimal-langpack-2.26-55.amzn2.aarch64
glibc-minimal-langpack-2.26-55.amzn2.x86_64**

**Comment by steve-chavez** [comment]
Have you tried using the binary compiled for Ubuntu ARM64? 

[postgrest-v8.0.0-ubuntu-aarch64.tar.xz](https://github.com/PostgREST/postgrest/releases/download/v8.0.0/postgrest-v8.0.0-ubuntu-aarch64.tar.xz)

Likely you'll need to install some packages for it to work.

---

Otherwise you'd need to compile from source. Basically:

```
git clone https://github.com/PostgREST/postgrest
cd postgrest
cabal update
cabal v2-install exe:postgrest --overwrite-policy=always
```

Then you can get the binary from `~/.cabal`.

**Comment by johndpope** [comment, marked as answer]
```
    __|  __|_  )
       _|  (     /   Amazon Linux 2 AMI
      ___|\___|___|

https://aws.amazon.com/amazon-linux-2/
4 package(s) needed for security, out of 19 available
Run "sudo yum update" to apply all updates.

➜  ~ cd postgrest
➜  postgrest git:(main) cabal update
Downloading the latest package list from hackage.haskell.org
➜  postgrest git:(main) cabal v2-install exe:postgrest --overwrite-policy=always
cabal: unrecognised command: v2-install (try --help)
```

UPDATE
cabal --version
cabal-install version 1.16.1.0
using version 1.16.0 of the Cabal library



➜  postgrest git:(main) curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | sh

Welcome to Haskell!

This script will download and install the following binaries:
  * ghcup - The Haskell toolchain installer
  * ghc   - The Glasgow Haskell Compiler
  * cabal - The Cabal build tool for managing Haskell software
  * stack - (optional) A cross-platform program for developing Haskell projects
  * hls   - (optional) A language server for developers to integrate with their editor/IDE

ghcup installs only into the following directory,
which can be removed anytime:
  /home/ec2-user/.ghcup

Press ENTER to proceed or ctrl-c to abort.
Note that this script can be re-run at any given time.

-------------------------------------------------------------------------------

Detected bash shell on your system...
Do you want ghcup to automatically add the required PATH variable to "/home/ec2-user/.bashrc"?

[P] Yes, prepend  [A] Yes, append  [N] No  [?] Help (default is "P").

A
-------------------------------------------------------------------------------
Do you want to install haskell-language-server (HLS)?
HLS is a language-server that provides IDE-like functionality
and can integrate with different editors, such as Vim, Emacs, VS Code, Atom, ...
Also see https://github.com/haskell/haskell-language-server/blob/master/README.md

[Y] Yes  [N] No  [?] Help (default is "N").

Y
-------------------------------------------------------------------------------
Do you want to install stack?
Stack is a haskell build tool similar to cabal that is used by some projects.
Also see https://docs.haskellstack.org/

[Y] Yes  [N] No  [?] Help (default is "N").

Y
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 44.4M  100 44.4M    0     0   114M      0 --:--:-- --:--:-- --:--:--  114M
/home/ec2-user/.ghcup/bin/ghcup: /lib64/libtinfo.so.6: no version information available (required by /home/ec2-user/.ghcup/bin/ghcup)
/home/ec2-user/.ghcup/bin/ghcup: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /home/ec2-user/.ghcup/bin/ghcup)
"_eghcup upgrade" failed!

**Reply by steve-chavez** [reply]
Try using one of the cabal downloads on https://www.haskell.org/cabal/download.html, I've never used ghcup, so I can't help much there.

**Reply by steve-chavez** [reply]
Use this one: https://downloads.haskell.org/~cabal/cabal-install-3.6.0.0/cabal-install-3.6.0.0-aarch64-linux-deb10.tar.xz

**Reply by steve-chavez** [reply]
Btw, stack might also work. See: https://postgrest.org/en/v8.0/install.html#building-from-source

**Reply by johndpope** [reply]
thanks for these links @steve-chavez  - I'll check it out early next week.

**Reply by johndpope** [reply]
no joy - 

➜  ~ tar -xJvf cabal-install-3.6.0.0-aarch64-linux-deb10.tar.xz
cabal
➜  ~ ls
cabal  cabal-install-3.6.0.0-aarch64-linux-deb10.tar.xz  fundfi  go  hosts  nginx.conf  ngrok  postgrest  postgrest-binary.tar.gz  PostGUI  quickstart  temp  tunnels
➜  ~ chmod +x cabal
➜  ~ ./cabal
./cabal: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by ./cabal)
➜  ~ curl -sSL https://get.haskellstack.org/ | sh

Detected Linux distribution: amzn

This installer doesn't support your Linux distribution, trying generic
bindist...

Sorry, currently only 64-bit (x86_64) Linux binary is available.


UPDATE
I'm looking at using glascow haskell compiler which has arm64 linux_aarch64 compatibility.
https://www.haskell.org/ghc/download_ghc_8_2_2.html#linux_aarch64

**Reply by steve-chavez** [reply]
> Sorry, currently only 64-bit (x86_64) Linux binary is available.

Hm, so stack is out then.

> I'm looking at using glascow haskell compiler which has arm64 linux_aarch64 compatibility.
https://www.haskell.org/ghc/download_ghc_8_2_2.html#linux_aarch64

Yes, try with direct downloads. Hope that works for you :crossed_fingers:

**Reply by johndpope** [reply]
UPDATE
that does actually build - but now I'm a bit stuck on how to use it. I've actually achived what I wanted using [www.supabase.io](supabase) in the cloud that handles (actually runs postgrest + and posgres) for me. I do want to run this stuff bare metal though.
You should hit them up for 💵💵💵💵
